authors/apps/profiles/views.py

```diff
@@ -0,0 +1,67 @@
+from rest_framework import generics, status
+from rest_framework.response import Response
+from rest_framework.filters import SearchFilter, OrderingFilter
+
+
+from .serializers import (ProfileSerializer)
+from .models import Profile
+
+
+class ViewAllProfiles(generics.ListAPIView):
+    """
+        this view will enable you to view all profiles in
+        the profiles table.
+    """
+    serializer_class = ProfileSerializer
+    filter_backends = (SearchFilter, OrderingFilter)
+    search_fields = ('bio', 'user__email', 'user__username')
+    ordering_fields = ('bio', 'user__username')
+    queryset = Profile.objects.all()
+
+
+class ProfileView(generics.RetrieveUpdateAPIView):
+    '''
+    class view to update or get a user profile
+    '''
+    serializer_class = ProfileSerializer
+    queryset = Profile.objects.all()
+    lookup_field = "username"
+
+    def get(self, request, username, *args, **kwargs):
+        """Retrieve a single profile"""
+        # if the username was a 2 or more strings then
+        # the username provided in the params
+        # would have a dash, we then check the list of user
+        # names to see the one that matches what we want exactly
+        # and pick the very first item which will match the item
+        username = username.replace("-", " ")
+        profile = Profile.objects.filter(
```

Comment:
Are you guys okay with how we handle updating and getting a specific user profile 
![image](https://user-images.githubusercontent.com/12128153/55378787-29804800-5523-11e9-9f07-119a224c3f76.png)

or can we have constraints on the username during account creation whereby the username cannot have spaces and capitalization.

To me this feels like a bit of a gamble since we are using `__icontains` and even though we are picking the very first item that has the best match its still not quite accurate.

For instance take a look at the screenshot below
![image](https://user-images.githubusercontent.com/12128153/55378975-effc0c80-5523-11e9-9b37-ce845dd18e1f.png)

Since we are using `__icontains` then it does give us the closest match to the username `tev` yet in essence there is no user by that username and the most accurate response would be `User with that username not found`

Reply:
I think we should limit what the username should be, like no uppercase letters or spaces, but numbers should be allowed to at least give them more options for the username.

Reply:
Would you consider adding a field on the Profile table and generate a unique slug/url from the username to be used to fetch the profile. you could use the profile id ..but to make the url human readable a slug is recommended
I would also recommend adding a created_at field on the profile table considering the updated at field will change on update

Reply:
Yeah,
You are right. I can make a username slug that we can use to get the profile from and also update the profile with. As for the created_at field, I was a bit reserved about it since the profile is created at the same time as a user and the user has that field already. 
If it was needed I could just add it to a serializer class and fetch it from the user object

Reply:
@martinMutuma @Kimaiyo077  thank you both for your input. This was eye opening

Reply:
@martinMutuma one more question, the username isn't saved in the profile. Its saved in the user model. If we were to generate a slug it would have to be in the user model. Instead of having that, why not just have constraints on the username like @Kimaiyo077  pointed out. It would be much easier for us and even for the user, considering also that the username is unique.

Reply:
A slug would be appropriate for an article because we don't necessarily require the title to be unique, but in our case the username is unique

Reply:
Names on sites are like (Martin mutuma ) username is like (Martin.mutuma or martinmutuma)
But for us we are using the same field for both so changing the functionality of the username would not be really solve your problem but create a request for changes later where users want their names to be separated

Reply:
Here's what I was thinking, a username is as the name suggests a name, not two names.
So What if we apply the constraints on the username and then in the profile, I add fields for 
1. Firstname
2. Lastname

Reply:
@martinMutuma  also take a look at the documentation.

When getting a user profile we use the `username` field not the slug

![image](https://user-images.githubusercontent.com/12128153/55386412-15931100-5538-11e9-8b4c-268b2185c750.png)

But for an article, we require a slug to get the article

![image](https://user-images.githubusercontent.com/12128153/55386446-28a5e100-5538-11e9-87d4-b34c240492e4.png)

Reply:
Totally agree with these. Get by username for profiles, then get by slug name for articles.